lms/services/canvas.py

```diff
@@ -1,36 +1,64 @@
 
 from lms import models
-from lms.services.exceptions import CanvasFileNotFoundInCourse
+from lms.services.exceptions import CanvasAPIPermissionError, CanvasFileNotFoundInCourse
 
 
 class CanvasService:
     """A high level Canvas service."""
 
     api = None
 
-    def __init__(self, canvas_api):
+    def __init__(self, canvas_api, assignment_service, file_service):
         self.api = canvas_api
-
-    def public_url_for_file(self, file_id, course_id, check_in_course=False):
-        """
-        Get a public URL for a Canvas file.
-
-        This will also attempt to check if the file is in the course to detect
-        course copy situations and fix it if we can.
-
-        :param file_id: The file to look up
-        :param course_id: The course the file should be in
-        :param check_in_course: Raise CanvasFileNotFoundInCourse if file_id isn't in
-            course_id
-        :return: A URL suitable for public presentation of the file
-        :raise  CanvasFileNotFoundInCourse: if check_in_course=True and file_id isn't in course_id
-            be in the course but isn't.
-        """
-
-        if check_in_course:
-            if not self.can_see_file_in_course(file_id, course_id):
-                raise CanvasFileNotFoundInCourse(file_id)
+        self._assignment_service = assignment_service
+        self._file_service = file_service
+
+    def public_url_for_file(
+        self,
+        application_instance,
+        resource_link_id,
+        file_id,
+        course_id,
+        check_in_course=False,
+    ):  # pylint:disable=too-many-arguments
+        mapped_file_id = self._assignment_service.get_canvas_mapped_file_id(
+            application_instance.tool_consumer_instance_guid, resource_link_id, file_id
+        )
+
+        # If there's a previously stored mapping for file_id use that instead.
+        effective_file_id = mapped_file_id or file_id
+
+        try:
+            return self._public_url(
+                effective_file_id,
+                course_id=course_id if check_in_course else None,
+            )
+        except (CanvasFileNotFoundInCourse, CanvasAPIPermissionError):
+            file_ = self._file_service.get(
+                application_instance, effective_file_id, type_="canvas_file"
+            )
+
+            if not file_:
+                raise
+
+            found_file_id = self.find_matching_file_in_course(course_id, file_)
+
+            if not found_file_id:
+                raise
+
+            self._assignment_service.set_canvas_mapped_file_id(
+                application_instance.tool_consumer_instance_guid,
+                resource_link_id,
+                file_id,
+                found_file_id,
+            )
+
+            return self._public_url(found_file_id)
+
+    def _public_url(self, file_id, course_id=None):
+        if course_id and not self.can_see_file_in_course(file_id, course_id):
+            raise CanvasFileNotFoundInCourse(file_id)
```

Comment:
I think this code could be a bit simpler (and avoid the `# pylint:disable=too-many-arguments`) if:

1. `FileService` didn't require the `application_instance` argument, instead getting the current request's `application_instance` injected by its factory
2. `AssignmentService.{get,set}_canvas_mapped_file_id()` took a single `module_item_configuration` argument and used `module_item_configuration.tool_consumer_instance_guid` and `module_item_configuration.resource_link_id`, instead of taking separate `tool_consumer_instance_guid` and `resource_link_id` arguments. Or alternatively if `{get,set}_canvas_mapped_file_id()` were `ModuleItemConfiguration` methods instead of `AssignmentService` methods

Then `public_url_for_file()` could look like this:

```python
    def public_url_for_file(
        self, module_item_configuration, file_id, course_id, check_in_course=False
    ):
        mapped_file_id = module_item_configuration.get_canvas_mapped_file_id(file_id)

        # If there's a previously stored mapping for file_id use that instead.
        effective_file_id = mapped_file_id or file_id

        try:
            return self._public_url(
                effective_file_id,
                course_id=course_id if check_in_course else None,
            )
        except (CanvasFileNotFoundInCourse, CanvasAPIPermissionError):
            file_ = self._file_service.get(effective_file_id, type_="canvas_file")

            if not file_:
                raise

            found_file_id = self.find_matching_file_in_course(course_id, file_)

            if not found_file_id:
                raise

            module_item_configuration.set_canvas_mapped_file_id(file_id, found_file_id)

            return self._public_url(found_file_id)
```